src/lazy/expanded/mod.rs

```diff
@@ -660,8 +780,9 @@ impl<'top, Encoding: Decoder> LazyExpandedValue<'top, Encoding> {
             ValueLiteral(value) => value.is_null(),
             Template(_, element) => element.value().is_null(),
             Constructed(_, value) => {
-                matches!(value, ExpandedValueRef::Null(_))
+                matches!(value, ValueRef::Null(_))
             }
+            EExp(eexp) => eexp.require_expansion_singleton().is_null(),
```

Comment:
🗺️ `is_null`, `has_annotations`, `annotations`, and `ion_type` can all be implemented without needing to actually evaluate the singleton macro. We remember these values from the compilation phase.